daemon/openrazer_daemon/hardware/device_base.py

```diff
@@ -104,13 +134,371 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                    pass
+
+                self.current_effect_speed = int(self.config[self._serial]['speed'])
+                self.current_wave_dir = int(self.config[self._serial]['wave_dir'])
+                effect_func_name = 'set' + self.current_effect[0].upper() + self.current_effect[1:]
+            else:
+                self.current_effect = 'spectrum'
+                effect_func_name = 'setSpectrum'
+
+            effect_func = getattr(self, effect_func_name, None)
+
+            if effect_func == None:
+                self.logger.info("%s: Invalid effect name %s; restoring to Spectrum.", self.__class__.__name__, effect_func_name)
+                effect_func_name = 'setSpectrum'
+                effect_func = getattr(self, effect_func_name, None)
+
+            if not effect_func == None:
+                if effect_func_name == 'setNone' or effect_func_name == 'setSpectrum' or effect_func_name == 'setBlinking' or effect_func_name == 'setBreathRandom':
+                    effect_func()
+                elif effect_func_name == 'setStatic' or effect_func_name == 'setBlinking' or effect_func_name == 'setBreathSingle':
+                    effect_func(self.current_effect_colors[0], self.current_effect_colors[1], self.current_effect_colors[2])
+                elif effect_func_name == 'setReactive':
+                    effect_func(self.current_effect_colors[0], self.current_effect_colors[1], self.current_effect_colors[2], self.current_effect_speed)
+                elif effect_func_name == 'setBreathDual':
+                    effect_func(self.current_effect_colors[0], self.current_effect_colors[1], self.current_effect_colors[2], self.current_effect_colors[3], self.current_effect_colors[4], self.current_effect_colors[5])
+                elif effect_func_name == 'setBreathTriple':
+                    effect_func(self.current_effect_colors[0], self.current_effect_colors[1], self.current_effect_colors[2], self.current_effect_colors[3], self.current_effect_colors[4], self.current_effect_colors[5], self.current_effect_colors[6], self.current_effect_colors[7], self.current_effect_colors[8])
+                elif effect_func_name == 'setWave':
+                    effect_func(self.current_wave_dir)
+                elif effect_func_name == 'setStarlightRandom':
+                    effect_func(self.current_effect_speed)
+                elif effect_func_name == 'setStarlightSingle':
+                    effect_func(self.current_effect_speed, self.current_effect_colors[0], self.current_effect_colors[1], self.current_effect_colors[2])
+                elif effect_func_name == 'setStarlightDual':
+                    effect_func(self.current_effect_speed, self.current_effect_colors[0], self.current_effect_colors[1], self.current_effect_colors[2], self.current_effect_colors[3], self.current_effect_colors[4], self.current_effect_colors[5])
+
+        # load last logo effect
+        if self.has_logo_effects:
```

Comment:
So much copy pasta :( It would be nice if you could try to unify those blocks into one that goes through 5 times with the different zones

Reply:
I never do nice things!
I will try using a map for the effect storage instead of independent variables to do so.

Reply:
Done through refactor